Flaky unit test `TestBlocksCleaner_ShouldRemoveBlocksOutsideRetentionPeriod`

Falked in this run https://github.com/grafana/mimir/runs/4883548679 but passed in the successive https://github.com/grafana/mimir/runs/4883702357

This assertion failed:

https://github.com/grafana/mimir/blob/303d683a5bb531315b7153ca6436d412acc6573c/pkg/compactor/blocks_cleaner_test.go#L657

```
--- FAIL: TestBlocksCleaner_ShouldRemoveBlocksOutsideRetentionPeriod (0.24s)
    blocks_cleaner_test.go:657: 
        	Error Trace:	blocks_cleaner_test.go:657
        	Error:      	Received unexpected error:
        	            	
        	            	metric output does not match expectation; want:
        	            	
        	            	# HELP cortex_bucket_blocks_count Total number of blocks in the bucket. Includes blocks marked for deletion, but not partial blocks.
        	            	# TYPE cortex_bucket_blocks_count gauge
        	            	cortex_bucket_blocks_count{user="user-1"} 1
        	            	cortex_bucket_blocks_count{user="user-2"} 0
        	            	# HELP cortex_bucket_blocks_marked_for_deletion_count Total number of blocks marked for deletion in the bucket.
        	            	# TYPE cortex_bucket_blocks_marked_for_deletion_count gauge
        	            	cortex_bucket_blocks_marked_for_deletion_count{user="user-1"} 0
        	            	cortex_bucket_blocks_marked_for_deletion_count{user="user-2"} 0
        	            	# HELP cortex_compactor_blocks_marked_for_deletion_total Total number of blocks marked for deletion in compactor.
        	            	# TYPE cortex_compactor_blocks_marked_for_deletion_total counter
        	            	cortex_compactor_blocks_marked_for_deletion_total{reason="retention"} 3
        	            	
        	            	got:
        	            	
        	            	# HELP cortex_bucket_blocks_count Total number of blocks in the bucket. Includes blocks marked for deletion, but not partial blocks.
        	            	# TYPE cortex_bucket_blocks_count gauge
        	            	cortex_bucket_blocks_count{user="user-1"} 1
        	            	cortex_bucket_blocks_count{user="user-2"} 1
        	            	# HELP cortex_bucket_blocks_marked_for_deletion_count Total number of blocks marked for deletion in the bucket.
        	            	# TYPE cortex_bucket_blocks_marked_for_deletion_count gauge
        	            	cortex_bucket_blocks_marked_for_deletion_count{user="user-1"} 0
        	            	cortex_bucket_blocks_marked_for_deletion_count{user="user-2"} 1
        	            	# HELP cortex_compactor_blocks_marked_for_deletion_total Total number of blocks marked for deletion in compactor.
        	            	# TYPE cortex_compactor_blocks_marked_for_deletion_total counter
        	            	cortex_compactor_blocks_marked_for_deletion_total{reason="retention"} 3
        	Test:       	TestBlocksCleaner_ShouldRemoveBlocksOutsideRetentionPeriod
```

--- a/pkg/compactor/blocks_cleaner.go
+++ b/pkg/compactor/blocks_cleaner.go
@@ -30,11 +30,16 @@ import (
 	util_log "github.com/grafana/mimir/pkg/util/log"
 )
 
+const (
+	defaultDeleteBlocksConcurrency = 16
+)
+
 type BlocksCleanerConfig struct {
-	DeletionDelay      time.Duration
-	CleanupInterval    time.Duration
-	CleanupConcurrency int
-	TenantCleanupDelay time.Duration // Delay before removing tenant deletion mark and "debug".
+	DeletionDelay           time.Duration
+	CleanupInterval         time.Duration
+	CleanupConcurrency      int
+	TenantCleanupDelay      time.Duration // Delay before removing tenant deletion mark and "debug".
+	DeleteBlocksConcurrency int
 }
 
 type BlocksCleaner struct {
@@ -360,8 +365,6 @@ func (c *BlocksCleaner) cleanUser(ctx context.Context, userID string) (returnErr
 	return nil
 }
 
-const deleteBlocksConcurrency = 16
-
 // Concurrently deletes blocks marked for deletion, and removes blocks from index.
 func (c *BlocksCleaner) deleteBlocksMarkedForDeletion(ctx context.Context, idx *bucketindex.Index, userBucket objstore.Bucket, userLogger log.Logger) {
 	blocksToDelete := make([]ulid.ULID, 0, len(idx.BlockDeletionMarks))
@@ -377,7 +380,7 @@ func (c *BlocksCleaner) deleteBlocksMarkedForDeletion(ctx context.Context, idx *
 	var mu sync.Mutex
 
 	// We don't want to return errors from our function, as that would stop ForEach loop early.
-	_ = concurrency.ForEachJob(ctx, len(blocksToDelete), deleteBlocksConcurrency, func(ctx context.Context, jobIdx int) error {
+	_ = concurrency.ForEachJob(ctx, len(blocksToDelete), c.cfg.DeleteBlocksConcurrency, func(ctx context.Context, jobIdx int) error {
 		blockID := blocksToDelete[jobIdx]
 
 		if err := block.Delete(ctx, userLogger, userBucket, blockID); err != nil {
@@ -415,7 +418,7 @@ func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials map
 	var mu sync.Mutex
 
 	// We don't want to return errors from our function, as that would stop ForEach loop early.
-	_ = concurrency.ForEachJob(ctx, len(blocks), deleteBlocksConcurrency, func(ctx context.Context, jobIdx int) error {
+	_ = concurrency.ForEachJob(ctx, len(blocks), c.cfg.DeleteBlocksConcurrency, func(ctx context.Context, jobIdx int) error {
 		blockID := blocks[jobIdx]
 
 		// We can safely delete only partial blocks with a deletion mark.

--- a/pkg/compactor/blocks_cleaner_test.go
+++ b/pkg/compactor/blocks_cleaner_test.go
@@ -96,10 +96,11 @@ func testBlocksCleanerWithOptions(t *testing.T, options testBlocksCleanerOptions
 	require.NoError(t, bucketClient.Upload(context.Background(), user4DebugMetaFile, strings.NewReader("some random content here")))
 
 	cfg := BlocksCleanerConfig{
-		DeletionDelay:      deletionDelay,
-		CleanupInterval:    time.Minute,
-		CleanupConcurrency: options.concurrency,
-		TenantCleanupDelay: options.tenantDeletionDelay,
+		DeletionDelay:           deletionDelay,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      options.concurrency,
+		TenantCleanupDelay:      options.tenantDeletionDelay,
+		DeleteBlocksConcurrency: 1,
 	}
 
 	reg := prometheus.NewPedanticRegistry()
@@ -231,9 +232,10 @@ func TestBlocksCleaner_ShouldContinueOnBlockDeletionFailure(t *testing.T) {
 	}
 
 	cfg := BlocksCleanerConfig{
-		DeletionDelay:      deletionDelay,
-		CleanupInterval:    time.Minute,
-		CleanupConcurrency: 1,
+		DeletionDelay:           deletionDelay,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      1,
+		DeleteBlocksConcurrency: 1,
 	}
 
 	logger := log.NewNopLogger()
@@ -290,9 +292,10 @@ func TestBlocksCleaner_ShouldRebuildBucketIndexOnCorruptedOne(t *testing.T) {
 	require.NoError(t, bucketClient.Upload(ctx, path.Join(userID, bucketindex.IndexCompressedFilename), strings.NewReader("invalid!}")))
 
 	cfg := BlocksCleanerConfig{
-		DeletionDelay:      deletionDelay,
-		CleanupInterval:    time.Minute,
-		CleanupConcurrency: 1,
+		DeletionDelay:           deletionDelay,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      1,
+		DeleteBlocksConcurrency: 1,
 	}
 
 	logger := log.NewNopLogger()
@@ -338,9 +341,10 @@ func TestBlocksCleaner_ShouldRemoveMetricsForTenantsNotBelongingAnymoreToTheShar
 	createTSDBBlock(t, bucketClient, "user-2", 30, 40, 2, nil)
 
 	cfg := BlocksCleanerConfig{
-		DeletionDelay:      time.Hour,
-		CleanupInterval:    time.Minute,
-		CleanupConcurrency: 1,
+		DeletionDelay:           time.Hour,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      1,
+		DeleteBlocksConcurrency: 1,
 	}
 
 	ctx := context.Background()
@@ -405,9 +409,10 @@ func TestBlocksCleaner_ShouldNotCleanupUserThatDoesntBelongToShardAnymore(t *tes
 	createTSDBBlock(t, bucketClient, "user-2", 20, 30, 2, nil)
 
 	cfg := BlocksCleanerConfig{
-		DeletionDelay:      time.Hour,
-		CleanupInterval:    time.Minute,
-		CleanupConcurrency: 1,
+		DeletionDelay:           time.Hour,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      1,
+		DeleteBlocksConcurrency: 1,
 	}
 
 	ctx := context.Background()
@@ -518,9 +523,10 @@ func TestBlocksCleaner_ShouldRemoveBlocksOutsideRetentionPeriod(t *testing.T) {
 	block4 := createTSDBBlock(t, bucketClient, "user-2", ts(-8), ts(-6), 2, nil)
 
 	cfg := BlocksCleanerConfig{
-		DeletionDelay:      time.Hour,
-		CleanupInterval:    time.Minute,
-		CleanupConcurrency: 1,
+		DeletionDelay:           time.Hour,
+		CleanupInterval:         time.Minute,
+		CleanupConcurrency:      1,
+		DeleteBlocksConcurrency: 1,
 	}
 
 	ctx := context.Background()

--- a/pkg/compactor/compactor.go
+++ b/pkg/compactor/compactor.go
@@ -454,10 +454,11 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
 
 	// Create the blocks cleaner (service).
 	c.blocksCleaner = NewBlocksCleaner(BlocksCleanerConfig{
-		DeletionDelay:      c.compactorCfg.DeletionDelay,
-		CleanupInterval:    util.DurationWithJitter(c.compactorCfg.CleanupInterval, 0.1),
-		CleanupConcurrency: c.compactorCfg.CleanupConcurrency,
-		TenantCleanupDelay: c.compactorCfg.TenantCleanupDelay,
+		DeletionDelay:           c.compactorCfg.DeletionDelay,
+		CleanupInterval:         util.DurationWithJitter(c.compactorCfg.CleanupInterval, 0.1),
+		CleanupConcurrency:      c.compactorCfg.CleanupConcurrency,
+		TenantCleanupDelay:      c.compactorCfg.TenantCleanupDelay,
+		DeleteBlocksConcurrency: defaultDeleteBlocksConcurrency,
 	}, c.bucketClient, c.shardingStrategy.blocksCleanerOwnUser, c.cfgProvider, c.parentLogger, c.registerer)
 
 	// Start blocks cleaner asynchronously, don't wait until initial cleanup is finished.

